### PR TITLE
Fix/BSA-202/Add base64 helper

### PR DESCRIPTION
**Pull request by shpran**
Relates to: [BSA-202](https://oat-sa.atlassian.net/browse/BSA-202)

**Changes:**
* Added base64 helper.

**Companion PRs:**
* [#418](https://github.com/oat-sa/extension-tao-item/pull/418)
* [#1499](https://github.com/oat-sa/extension-tao-itemqti/pull/1499)
* [#81](https://github.com/oat-sa/extension-tao-testqti-previewer/pull/81)

### CODE REVIEW

File: manifest.php [hunk 1]
```diff
@@ -62,7 +62,7 @@
     'label' => 'TAO Base',
     'description' => 'TAO meta-extension',
     'license' => 'GPL-2.0',
-    'version' => '45.1.2',
+    'version' => '45.1.3',
```
Comment (wazelin) [inline manifest.php:65]: This looks like a new feature to me, kindly consider bumping a minor version instead.
```suggestion
    'version' => '45.2.0',
```

File: helpers/Base64.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+{
+    /**
+     * @param $data
+     *
+     * @return bool
+     */
+    public static function isEncodedImage($data): bool
+    {
+        return is_string($data) && preg_match('/^(data:.*;base64)/', $data) && getimagesize($data) !== false;
+    }
```
Comment (wazelin) [inline helpers/Base64.php:20]: Consider adding a less strict `isEncoded()` method, just to complete the helper.

```suggestion
    public function isEncoded($data): bool
    {
        return is_string($data) && preg_match('/^data:.*;base64/', $data)
    }

    /**
     * @param $data
     *
     * @return bool
     */
    public static function isEncodedImage($data): bool
    {
        return self::isEncoded($data) && getimagesize($data) !== false;
    }
```

You don't need a capturing group in the regex BTW.

**Comment by wazelin** [comment]
Nice touch with extracting the method into a `core` helper!

Would you consider adding a few simple test cases for it?

File: helpers/Base64.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * @package oat\tao\helpers
+ */
+class Base64
+{
+    /**
+     * @param $data
+     *
+     * @return bool
+     */
+    public function isEncoded($data): bool
```
Comment (wazelin) [inline helpers/Base64.php:17]: My apologies, I forgot to add a `static` keyword in my original suggestion.
```suggestion
    public static function isEncoded($data): bool
```

File: test/unit/helpers/Base64Test.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * @dataProvider dataProvider
+     *
+     * @param $value
+     * @param bool $result
+     */
+    public function testBase64EncodingImage($value, bool $result): void
+    {
+        $this->assertEquals(Base64::isEncodedImage($value), $result);
```
Comment (wazelin) [inline test/unit/helpers/Base64Test.php:57]: `assertEquals()` accepts an expected value as a first argument.

```suggestion
        $this->assertEquals($result, Base64::isEncoded($value));
    }

    /**
     * @dataProvider dataProvider
     *
     * @param $value
     * @param bool $result
     */
    public function testBase64EncodingImage($value, bool $result): void
    {
        $this->assertEquals($result, Base64::isEncodedImage($value));
```

**Comment by wazelin** [comment]
GJ!
Just a very minor quibble to address, and the PR is good to go.

File: test/unit/helpers/Base64Test.php (hunk 4 above)
Comment (shpran) [reply test/unit/helpers/Base64Test.php:57]: You are right! Thanks for your comment! :)

File: test/unit/helpers/Base64Test.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            ],
+            'string' => [
+                'value' => 'string',
+                'result' => false,
+            ],
+            'encodedString' => [
+                'value' => self::ENCODED_IMAGE,
+                'result' => true,
+            ],
+            'encodedImage' => [
```
Comment (ktsirkou) [inline test/unit/helpers/Base64Test.php:78]: What is the difference between the last 2 cases? 
I would like to add a case where is encoded string but not image, wdyt?

Comment (shpran) [reply test/unit/helpers/Base64Test.php:104]: Thanks for your comment :)
Done.

**Pull request merged**